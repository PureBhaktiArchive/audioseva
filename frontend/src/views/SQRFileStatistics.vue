<template>
  <div>
    <header>
      <h1>SQR Statistics</h1>
    </header>
    <files-by-status
      :isLoading="isLoadingFiles"
      :countByStatus="fileCountByStatus"
      :items="filesByStatus"
    >
    </files-by-status>
    <done-statistics :doneStatistics="doneStatistics"></done-statistics>
    <spare-by-language :spareByLanguage="spareByLanguage"></spare-by-language>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import _ from "lodash";
import fb from "@/firebaseApp";
import {
  ISQRFileVueFire,
  ISpareByLanguage,
  IFileByStatus,
  ICount
} from "@/types/DataTable";
import FilesByStatus from "@/components/SQRStatistics/FilesByStatus.vue";
import DoneStatistics from "@/components/SQRStatistics/DoneStatistics.vue";
import SpareByLanguage from "@/components/SQRStatistics/SpareByLanguage.vue";

@Component({
  name: "SQRFileStatistics",
  components: { DoneStatistics, FilesByStatus, SpareByLanguage }
})
export default class SQRFileStatistics extends Vue {
  lists!: { [key: string]: { [key: string]: ISQRFileVueFire } };
  isLoadingFiles: boolean = false;
  doneFiles: any = null;
  doneStatistics: ICount = {};
  filesByStatus: IFileByStatus[] = [];
  spareByLanguage: ISpareByLanguage = {};
  fileCountByStatus: ICount = {};
  headers = [
    { text: "File name", value: "filename" },
    { text: "Status", value: "status" },
    { text: "Languages", value: "languages" }
  ];

  mounted() {
    this.fetchLists();
  }

  fetchLists() {
    this.isLoadingFiles = true;
    const date = new Date();
    this.$bindAsObject(
      "lists",
      fb.database().ref("files"),
      null,
      this.extractFiles
    );
    this.$bindAsArray(
      "doneFiles",
      fb
        .database()
        .ref("sqr/submissions")
        .orderByChild("completed")
        .startAt(date.setDate(date.getDate() - 5)),
      null,
      this.doneFileStatistics
    );
  }

  doneFileStatistics() {
    const doneStatistics = {};
    _.forEach(this.doneFiles, file => {
      const date = this.getDate(new Date(file.completed));
      _.set(doneStatistics, date, _.get(doneStatistics, date, 0) + 1);
    });
    this.doneStatistics = doneStatistics;
  }

  getDate(date: Date) {
    const today = new Date();
    const dateString = date.toDateString().split(" ");
    return today.toDateString() === date.toDateString()
      ? "today"
      : `${dateString[1]} ${dateString[2]}`;
  }

  extractFiles() {
    const statusByList = {};
    const spareByLanguage = {};
    const fileCountByStatus = {};
    _.forIn(this.lists, (list, listName) => {
      if (listName !== ".key") {
        const listStatusTotal = `${listName}.GRAND`;
        // file count by list
        _.set(statusByList, listStatusTotal, Object.keys(list).length);

        _.forIn(
          list,
          ({
            soundQualityReporting: { status } = { status: "Spare" },
            languages
          }) => {
            const listStatus = `${listName}.${status}`;

            // spare files by language
            if (status === "Spare") {
              _.forEach(languages, (language: string) => {
                _.set(
                  spareByLanguage,
                  language,
                  _.get(spareByLanguage, language, 0) + 1
                );
              });
            }
            // total file count per status, used in footer for GRAND
            _.set(
              fileCountByStatus,
              status,
              _.get(fileCountByStatus, status as string, 0) + 1
            );
            // total file count
            _.set(
              fileCountByStatus,
              "GRAND",
              _.get(fileCountByStatus, "GRAND", 0) + 1
            );
            // file count per list per status
            _.set(
              statusByList,
              listStatus,
              _.get(statusByList, listStatus, 0) + 1
            );
          }
        );
      }
    });
    this.fileCountByStatus = fileCountByStatus;
    this.filesByStatus = Object.entries(statusByList).map(([list, stats]) => ({
      ...stats,
      list
    }));
    this.spareByLanguage = spareByLanguage;
    this.isLoadingFiles = false;
  }
}
</script>

<style scoped>
</style>
