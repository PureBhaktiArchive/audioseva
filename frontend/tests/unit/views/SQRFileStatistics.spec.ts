import { shallowMount } from "@vue/test-utils";
import SQRFileStatistics from "@/views/SQRFileStatistics.vue";

const lists = {
  list1: {
    file1: {
      status: "Spare",
      languages: ["English"]
    },
    file2: {
      status: "Spare",
      languages: ["English"]
    }
  },
  list2: {
    file3: {
      status: "Given",
      languages: ["English"]
    },
    file4: {
      status: "Spare",
      languages: ["Bengali"]
    }
  }
};

describe('SQRFileStatistics', () => {

  it('should extract files', () => {
    const wrapper = shallowMount(SQRFileStatistics, {
      mocks: {
        lists
      },
      methods: {
        fetchLists: () => null
      }
    });
    wrapper.vm.extractFiles();
    expect(wrapper.vm.filesByStatus).toMatchSnapshot("files by status");
    expect(wrapper.vm.spareByLanguage).toMatchSnapshot("spare by language");
    expect(wrapper.vm.fileCountByStatus).toMatchSnapshot("file count by status");
  });

  it("should get doneStatistics", () => {
    const today = new Date(1540685939096);
    const doneFiles = [
      {
        [".key"]: "list",
        completed: new Date(today).setDate(today.getDate() - 5)
      },
      {
        [".key"]: "list1",
        completed: new Date(today).setDate(today.getDate() - 4)
      },
      {
        [".key"]: "list2",
        completed: new Date(today).setDate(today.getDate() - 3)
      },
      {
        [".key"]: "list3",
        completed: new Date(today).setDate(today.getDate() - 2)
      },
      {
        [".key"]: "list4",
        completed: today
      },
      {
        [".key"]: "list5",
        completed: today
      }
    ];
    const oldDate = Date;
    // @ts-ignore
    global.Date = class extends oldDate {
      constructor(...args: any[]) {
        super();
        if (args.length) {
          // @ts-ignore
          return new oldDate(...args);
        }
        return new oldDate(1540685939096);
      }
    };
    const wrapper = shallowMount(SQRFileStatistics, {
      mocks: {
        doneFiles
      },
      methods: {
        fetchLists: () => null
      }
    });
    wrapper.vm.doneFileStatistics();
    expect(wrapper.vm.doneStatistics).toMatchSnapshot();
    // @ts-ignore
    global.Date = oldDate;
  });
});
