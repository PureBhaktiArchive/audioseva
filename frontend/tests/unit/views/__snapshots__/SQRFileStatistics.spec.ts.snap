// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQRFileStatistics should extract files: file count by status 1`] = `
Object {
  "GRAND": 4,
  "Given": 1,
  "Spare": 3,
}
`;

exports[`SQRFileStatistics should extract files: files by status 1`] = `
Array [
  Object {
    "GRAND": 2,
    "Spare": 2,
    "list": "list1",
  },
  Object {
    "GRAND": 2,
    "Given": 1,
    "Spare": 1,
    "list": "list2",
  },
]
`;

exports[`SQRFileStatistics should extract files: spare by language 1`] = `
Object {
  "Bengali": 1,
  "English": 2,
}
`;

exports[`SQRFileStatistics should get doneStatistics 1`] = `
Object {
  "Oct 22": 1,
  "Oct 23": 1,
  "Oct 24": 1,
  "Oct 25": 1,
  "today": 2,
}
`;
